### PR TITLE
Filter and sorting for product page

### PR DESCRIPTION
**Pull request by samyakshah3008**
Here is the filters and sorting functionality for e commerce Triple S Store 

Features: 
- Price ranger 
- Include or exclude stock products
- Sort by category
- Ratings
- Price filter: High to low or low to high

Here is the live preview link: https://graceful-otter-7b5d2c.netlify.app/

### CODE REVIEW

File: src/components/SideBar.jsx [hunk 1]
```diff
@@ -1,4 +1,14 @@
+import { useFilter } from "../contexts/filtercontext";
+import { useProduct } from "../contexts/product-context";
+
 export function SideBar() {
+  const { state, dispatch } = useFilter();
+  const { products } = useProduct();
+
+  // console.log(state)
```
Comment (manojasarsa) [inline src/components/SideBar.jsx:8]: you can remove these console.log

File: src/index.js [hunk 2]
```diff
@@ -4,16 +4,19 @@ import App from "./App";
...
+import { FilterProvider } from "./contexts/filtercontext";
 
 // Call make Server
 makeServer();
 
 ReactDOM.render(
   <React.StrictMode>
     <BrowserRouter>
+      <FilterProvider>
       <ProductProvider>
```
Comment (manojasarsa) [inline src/index.js:16]: You can give some identation here, to make it look good.

**Review by manojasarsa — APPROVED** [review]
Rest the code looks and organized.
Good Work, Samyak 🥇

**Review by shrey27 — APPROVED** [review]
Your overall UI looks good. But on products page, i can see your checkbox and labels are not aligned properly. So fix that
Code is overall fine, but I would suggest use array or objects with help of reducer to handle category filter otherwise code will become repetitive and un-scalable
I would suggest work on the issues before merging

File: src/reducers/filter-reducer.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    case "FAST_DELIVERY":
+      return { ...state, fastDelivery: action.payload };
+
+    case "SORT_BY":
+      return { ...state, sortBy: action.payload };
+
+    case "RATINGS":
+      return { ...state, ratings: action.payload };
+
+    case "MRF":
```
Comment (shrey27) [inline src/reducers/filter-reducer.js:18]: make a single use case for category and append or remove or toggle the state using payload.

File: src/backend/db/products.js [hunk 4]
```diff
@@ -18,59 +24,116 @@ export const products = [
...
     originalPrice: "1000",
-    price: "499",
-    categoryName: "fiction",
+    price: 1500,
+    categoryName: "nike",
     percentOff: "53%",
+    inStock: true,
+    fastDelivery: false,
+    ratings: 2
   },
```
Comment (hinanshisuthar) [inline src/backend/db/products.js:60]: Would suggest you to format your code using any code formatter like prettier.

File: src/components/Card.jsx [hunk 5]
```diff
@@ -19,6 +19,9 @@ const Card = ({ item }) => {
             </div>
           </div>
           <div className="discount-percent"> {item.percentOff} </div>
+          <div> {item.inStock ? "Available in stock" : "Not available"} </div>
```
Comment (hinanshisuthar) [inline src/components/Card.jsx:22]: Good use of ternary operator

File: src/components/SideBar.jsx [hunk 6]
```diff
@@ -12,59 +22,102 @@ export function SideBar() {
...
+              <label>Include only in stock products</label>
             </div>
             <div>
-            <input className="fastdelivery" type = "radio"/>
-            <label>Fast delivery</label>
+              <input
+                className="fastdelivery"
+                type="checkbox"
+                onClick={(e) =>
+                  dispatch({ type: "FAST_DELIVERY", payload: e.target.checked })
```
Comment (hinanshisuthar) [inline src/components/SideBar.jsx:54]: Got to know something like e.target.checked exists.

**Review by hinanshisuthar — COMMENTED** [review]
Suggestions given. Overall your code looks good. Your can proceed after making these small changes. If you have time later on do refactor it as the code seems repetitive. But as of now, good work Samyak!

**Pull request merged**